### PR TITLE
Implementation of GatewayConfig

### LINKED ISSUE #166: Implement TokenConfig
- Update WETH config for Goerli chain.

- Update WETH config for dev chain.

- Add method `readConfigFromFile` which will read config from file, validate it and return output as JSON object.

- Create class TokenConfig. Proposed interface:
     ```
      export default class TokenConfig {
                mosaicConfig: MosaicConfig
                auxChainId: Integer
                originContracts: OriginContracts
                auxiliaryContracts: AuxiliaryContracts
            }
      export class OriginContracts {}
      export class AuxiliaryContracts {}
      ```
- OriginContracts class is interface for origin chain contract addresses.

- AuxiliaryContracts class is interface for auxiliary chain contract addresses.   

- Implementor can decide if existing interfaces of Mosaic Config for originContracts, 
   auxiliaryContracts can be reused or not.    

- Token Config should be verified.

- TokenConfig.schema.json should be created to validate token config

- JSON validator should contain validations for mandatory and optional parameters.

- MosaicConfig.schema.json can be used for reference.

- As per discussion TokenConfig is renamed to GatewayConfig.
  Reference: #180 

Fixes #166 #180

### PR DESCRIPTION
**Pull request by abhayks1**
PR covers below:

- Updated WETH.json config for Goerli chain.

- Updated WETH.json config for dev chain.

- Added method `fromFile` which will read config from file, validate it and return output as JSON object.

- Create class <strike>TokenConfig</strike> GatewayConfig with interface:
     ```
      export default class GatewayConfig {
                mosaicConfigFilePath: MosaicConfig
                auxChainId: Integer
                originContracts: OriginContracts
                auxiliaryContracts: AuxiliaryContracts
            }
      export class OriginContracts {}
      export class AuxiliaryContracts {}
      ```

- OriginContracts class holds origin chain contract addresses.

- AuxiliaryContracts class holds auxiliary chain contract addresses.   

- <strike>TokenConfig.schema.json</strike> GatewayConfig.schema.json is created to validate gateway config. It contains validations for mandatory and optional parameters.

Fixes #166

### CODE REVIEW

File: chains/goerli/WETH.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  },
+  "auxiliaryContracts":{
+    "coGatewayOrganizationAddress":"0x47Ef953C4372457754d3471Ca2BB1b5d6AD8F245",
+    "eip20CoGatewayAddress":"0x962e1404D6957d8fF56F8fE79a3fC2B670C3d578",
+    "utilityTokenAddress":"0xBB5676d85d28DA039F982C07E4217fB0FDB2c2ef",
+    "gatewayLibAddress":"0xa718f65c96a177d8967E8f6D67A7872b26771495",
+    "messageBusAddress":"0x8F3578462BaF62e48Ad02820f4b7cd9788B200B6",
+    "merklePatriciaLibAddress":"0x4473FBF3390e49e79b27D229b9BF230DAD78414e"
+  }
+}
```
Comment (sarvesh-ost) [inline chains/goerli/WETH.json]: This file should in aux chain directory.

File: src/Config/TokenConfig.schema.json [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      ],
+      "title": "Auxiliary",
+      "description": "Contract addresses of auxiliary chain"
+    },
+    "TokenConfig": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "mosaicConfigFilePath": {
+          "type": "string",
```
Comment (sarvesh-ost) [inline src/Config/TokenConfig.schema.json]: Can you write a regex here?

File: src/Config/TokenConfig.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
+import MosaicConfig, { Address } from "./MosaicConfig";
+import {
+  InvalidTokenConfigException,
+  TokenConfigNotFoundException
+} from "../Exception";
+import * as fs from "fs-extra";
+import {Validator} from "jsonschema";
+const schema = require('./TokenConfig.schema.json');
+const path = require('path');
```
Comment (sarvesh-ost) [inline src/Config/TokenConfig.ts]: Please fix the linter error.

File: src/Config/TokenConfig.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+   * @param {string} filePath tokenConfig absolute path.
+   *
+   * @return {TokenConfig} TokenConfig object.
+   */
+  public static fromFile(filePath: string): TokenConfig {
+    if (fs.existsSync(filePath)) {
+      const configObject = TokenConfig.readConfigFromFile(filePath);
+      return new TokenConfig(configObject);
+    }
+    throw new TokenConfigNotFoundException(`Missing config file at path: ${filePath}`);
```
Comment (sarvesh-ost) [inline src/Config/TokenConfig.ts]:
```suggestion
    throw new TokenConfigNotFoundException(`Missing token config file at path: ${filePath}`);
```

File: src/Config/TokenConfig.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+   *
+   * @return {object} Json parsed object.
+   */
+  private static readConfigFromFile(filePath: string): object {
+    const configString = fs.readFileSync(filePath).toString();
+    if (configString && configString.length > 0) {
+      const configObject = JSON.parse(configString);
+      TokenConfig.validateSchema(configObject);
+      configObject.mosaicConfigFilePath = TokenConfig.resolveHomePath(
+        configObject.mosaicConfigFilePath
```
Comment (sarvesh-ost) [inline src/Config/TokenConfig.ts]:
```suggestion
        configObject.mosaicConfigFilePath,
```

File: src/Config/TokenConfig.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    throw new InvalidTokenConfigException(`blank config file found at: ${filePath}`);
+  }
+
+  /**
+   * It resolves token config file path to absolute path after resolving tilde.
+   *
+   * @param filePath Token config file path. e.g. ~/.mosaic/goerli/WETH.json
+   * @return {string} Absolute file path.
+   */
+  private static resolveHomePath(filePath: string) {
```
Comment (sarvesh-ost) [inline src/Config/TokenConfig.ts]: This logic should be moved to other class may be `FileSystem` or `Directory`

File: src/Config/TokenConfig.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  /**
+   * Read config from file, validate it and return as JSON object.
+   *
+   * @param {string} filePath tokenConfig absolute path.
+   *
+   * @return {object} Json parsed object.
+   */
+  private static readConfigFromFile(filePath: string): object {
```
Comment (sarvesh-ost) [inline src/Config/TokenConfig.ts]: Do we need method to read token config based on chain identifier?

File: chains/goerli/WETH.json (hunk 1 above)
Comment (abhayks1) [reply chains/goerli/WETH.json]: Done!

File: src/Config/TokenConfig.ts (hunk 5 above)
Comment (abhayks1) [reply src/Config/TokenConfig.ts]: Done!

File: src/Config/TokenConfig.ts (hunk 6 above)
Comment (abhayks1) [reply src/Config/TokenConfig.ts]: Done!

File: src/Config/TokenConfig.ts (hunk 7 above)
Comment (abhayks1) [reply src/Config/TokenConfig.ts]: I don't think so there is use case right now. Facilitator will read from tokenConfig file path.
We can add whenever the use case arises.

File: src/Config/TokenConfig.ts (hunk 4 above)
Comment (abhayks1) [reply src/Config/TokenConfig.ts]: Done!

File: src/Config/TokenConfig.schema.json (hunk 2 above)
Comment (abhayks1) [reply src/Config/TokenConfig.schema.json]: Added!

File: src/Config/TokenConfig.ts (hunk 7 above)
Comment (abhayks1) [reply src/Config/TokenConfig.ts]: Added TokenConfig.fromChain method.

**Comment by abhayks1** [comment]
Ready for review again...

File: chains/goerli/WETH.json (hunk 1 above)
Comment (sarvesh-ost) [reply chains/goerli/WETH.json]: File name should be <gatewayaddress>.json?

File: docker/0xae02c7b1c324a8d94a564bc8d713df89eae441fe.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{
+  "mosaicConfigFilePath":"~/.mosaic/dev/mosaic.json",
```
Comment (sarvesh-ost) [inline docker/0xae02c7b1c324a8d94a564bc8d713df89eae441fe.json]:
```suggestion
  "mosaicConfigFilePath":"~/.mosaic/dev-origin/mosaic.json",
```

File: src/Config/GatewayConfig.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  /**
+   * Construct GatewayConfig object
+   *
+   * @param originChain Origin chain identifier.
+   * @param auxChainId Auxiliary chain Id.
+   * @param gatewayAddress Address of Gateway.
+   *
+   * @return mosaic config
```
Comment (sarvesh-ost) [inline src/Config/GatewayConfig.ts]:
```suggestion
   * @return gateway config
```

File: src/Config/GatewayConfig.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+  public mosaicConfig: MosaicConfig;
+
+  public auxChainId: number;
+
+  public originContracts: OriginContracts;
+
+  public auxiliaryContracts: AuxiliaryContracts;
+
+  private constructor(config: any) {
+    this.mosaicConfig = MosaicConfig.fromFile(config.mosaicConfigFilePath);
```
Comment (sarvesh-ost) [inline src/Config/GatewayConfig.ts]: will this fail if file doesn't exist?

File: src/Directory.ts [hunk 11]
```diff
@@ -164,4 +164,23 @@ export default class Directory {
...
   }
+
+  /**
+   * Returns the full path of GatewayConfig for a given origin, auxiliary and gatewayAddress.
+   *
+   * @param originChain Origin chain identifier.
+   * @param auxChainId Auxiliary chain Id.
+   * @param gatewayAddress Address of Gateway.
+   *
+   * @return Path of mosaic config file.
```
Comment (sarvesh-ost) [inline src/Directory.ts]:
```suggestion
   * @return Path of gateway config file.
```

File: docker/0xae02c7b1c324a8d94a564bc8d713df89eae441fe.json (hunk 8 above)
Comment (abhayks1) [reply docker/0xae02c7b1c324a8d94a564bc8d713df89eae441fe.json]: why dev-origin and not dev?

File: src/Config/GatewayConfig.ts (hunk 10 above)
Comment (abhayks1) [reply src/Config/GatewayConfig.ts]: Yes

File: docker/0xae02c7b1c324a8d94a564bc8d713df89eae441fe.json (hunk 8 above)
Comment (sarvesh-ost) [reply docker/0xae02c7b1c324a8d94a564bc8d713df89eae441fe.json]: Because mosaic config is stores in `~/.mosaic/dev-origin` directory.

**Pull request merged**